### ISSUE TITLE
Link to the reason if a package is not available for an arch

### ISSUE THREAD

**Issue by probonopd** [FreshPorts/freshports#384]
It would be enormously helpful if instead of showing empty boxes on

https://www.freshports.org/print/hplip

![image](https://user-images.githubusercontent.com/2480569/189477251-5e410325-8ff4-4b3f-92ff-c22bc00e89d2.png)

these empty boxes would contain links to

![image](https://user-images.githubusercontent.com/2480569/190873772-3d4f133a-8574-4c03-b966-59f49ca902a1.png)

instead so that one could know why the package is not available for this arch.

**Comment by dlangille** [comment]
Do you have an example URL please?

**Comment by probonopd** [comment]
Stll trying to figuring this out myself... https://github.com/helloSystem/hello/discussions/382

**Comment by probonopd** [comment]
Maybe link to https://portsfallout.com/port?port=^print/hplip$ (although it doesn't show any failure at the moment, even though the package is missing; @dbaio do you know why?)

**Comment by grahamperrin** [comment]
Cross-reference: 

* https://github.com/helloSystem/hello/discussions/384#discussioncomment-3728654

<http://beefy18.nyi.freebsd.org/build.html?mastername=main-amd64-default&build=p47ff75a1891e_s2c2ef670a7> `print/hplip` was _skipped_, not a failure: 

![image](https://user-images.githubusercontent.com/192271/192164024-2fc9f256-3e34-4012-87de-eab930587a2a.png)

* the **reason** for failure was `net-mgmt/net-snmp`.

<https://portsfallout.com/port?port=net-mgmt%2Fnet-snmp>

**Comment by dlangille** [comment]
The portsfallout.out links where added via #181 about two years ago.

**Comment by grahamperrin** [comment]
@probonopd https://www.freshports.org/faq.php#fallout

**Comment by probonopd** [comment]
What I mean is to put a link into the otherwise empty boxes, maybe somewhat like this:

![image](https://user-images.githubusercontent.com/2480569/192165502-27d18870-41a3-458b-8ed0-fa1299130ddc.png)

**Comment by dlangille** [comment]
See https://dev.freshports.org/archivers/arc/#packages

![Screen Shot 2022-09-25 at 6 01 19 PM](https://user-images.githubusercontent.com/1626436/192167583-34ae5fa6-0aba-4c0c-b65e-ea986b90d3ee.png)

**Comment by grahamperrin** [comment]
> … Fallout would have to become a bit more intelligent, so that it would tell us that `print/hplip` was skipped because `net-mgmt/net-snmp` failed.

From <https://portsfallout.com/about>: 

>> … an easy way to search/query the [FreeBSD pkg-fallout](https://lists.freebsd.org/archives/freebsd-pkg-fallout/) archive. …

https://github.com/dbaio/portsfallout#readme

https://github.com/dbaio/portsfallout/blob/f65703382a2de6bb1456e36b5a646fbb0dbac619/scripts/pkgfallout_scrapy_spider.py#L25

This month's archived e-mails about `net-mgmt/net-snmp` include <https://lists.freebsd.org/archives/freebsd-pkg-fallout/2022-September/275174.html>, _[package - main-amd64-default][net-mgmt/net-snmp] Failed for net-snmp-5.9.1_1,1 in build_.

<https://lists.freebsd.org/> there's nothing like a list for skipped ports.

**Comment by dlangille** [comment]
> https://lists.freebsd.org/ there's nothing like a list for skipped ports.

Based on that: nothing more to do here.

**Comment by grahamperrin** [comment]
<https://github.com/FreshPorts/freshports/issues/384#issuecomment-1257279011>

> … maybe somewhat like this:
> 
> ![image](https://user-images.githubusercontent.com/2480569/192165502-27d18870-41a3-458b-8ed0-fa1299130ddc.png)
> 
> …

@dlangille please, it is possible for each single fallout icon to appear only if the user points at the cell? 

Maybe an empty cell until pointed at, although (second screenshot below) the previously used dash/hyphen was also effective as a hint.

### Rationale

The **masses of hazardous fallout** (non-availability) sometimes detract from what's of greater interest to the majority of users: 

* availability. 

Compare, for example: 

![fallout galore](https://user-images.githubusercontent.com/192271/193460643-2a21cfe9-f2bc-432b-8dd2-a8b17ba58544.png)

– with this, from a Wayback Machine capture a few months ago: 

![easier to see what's available](https://user-images.githubusercontent.com/192271/193460672-46cb554c-485e-45c1-af66-f716eec4b148.png)

**Comment by probonopd** [comment]
Agree, a "hover" showing the icon only when the mouse is in the cell would also do the job.

**Comment by dlangille** [comment]
These changes are now on test, but not the hover issue mentioned above.  Please review them and see if they are OK.

Then I'll tackle the hover.

**Comment by grahamperrin** [comment]
1. <https://pkg-status.freebsd.org/builds?jailname=131amd64&type=package>
2. <http://beefy14.nyi.freebsd.org/build.html?mastername=131amd64-quarterly&build=07882fc18e19> for quarterly 
3. `firebird30-client-3.0.7_2` amongst the failures
4. open _coredump_ in a new background tab
5. under _origin_, open _databases/firebird30-client_
6. edit `www` to `test` then add `#packages` to the tail of `https://test.freshports.org/databases/firebird30-client/`
7. at <https://test.freshports.org/databases/firebird30-client/#packages> <strike>click </strike>point at, but do not click, the quarterly (second) fallout icon for `FreeBSD:13:amd64`
8. from what's in the pop-up, memorise the build date
9. use <strike>the quarterly (second) fallout icon for `FreeBSD:13:amd64` </strike>any fallout icon to open <https://portsfallout.com/fallout?port=databases%2Ffirebird30-client%24> in a new foreground tab
10. **ignore** the uppermost log for _131amd64-quarterly_
11. click the log for _131amd64-quarterly_ that matches the **memorised** build date
12. bring the other new tab (from _coredump_) to front
13. check the URLs of the two tabs. 

The two are effectively the same: 

* `http://beefy14.nyi.freebsd.org/data/131amd64-quarterly/07882fc18e19/logs/errors/firebird30-client-3.0.7_2.log` via the _coredump_ link
* `http://beefy14.nyi.freebsd.org/data/131amd64-quarterly/07882fc18e19/logs/firebird30-client-3.0.7_2.log` – without `errors/` in the URL – via portsfallout.com

– and at <https://lists.freebsd.org/archives/freebsd-pkg-fallout/2022-October/date.html> by date for this month, the penultimate match for _131amd64-quarterly][databases/firebird30-client] Failed for firebird30-client_ is <https://lists.freebsd.org/archives/freebsd-pkg-fallout/2022-October/279283.html>, Sun, 09 Oct 2022 01:34:57 UTC.

Test result: OK. **However** …

**Comment by grahamperrin** [comment]
### UX

… at <https://test.freshports.org/databases/firebird30-client/#packages> there's an implicit requirement to pay careful attention to the build date, if the end user is to identify the correct log file. 

Let's be realistic about the likelihood of users actually paying attention and then knowing what to do with the date. 

What's most obvious (uppermost) at the fallout list might be **not** an explanation for a gap in the table of packages at FreshPorts. 

----

With the databases/firebird30-client example above, at the time of writing there's a parallel build at <https://pkg-status.freebsd.org/beefy14/> in progress (neither done, nor crashed), it's this build that produced the log for 131amd64-quarterly that's currently uppermost at the fallout list.

**Comment by grahamperrin** [comment]
<https://test.freshports.org/graphics/py-cairo/> now has 275 fallout links to a single page that confirms, no fallout. 

----

For a port that truly does have fallout: <https://test.freshports.org/databases/firebird30-client/#packages> **in theory**, you might customise each fallout cell within the table, to _precisely_ directly link to the historic reason for fallout (or skip) at that point in time. 

**In reality**, for now, the more that I think about this, the _general_ direction to the (one) fallout page for a port is better served by the (one) existing fallout icon that's already at the head of production service (www). 

----

Sorry … that's not intended to be defeatist. Just realistic about what's gained if multiple fallout links to a single page are added. (Am I missing something?)

**Comment by dlangille** [comment]
> 1. https://pkg-status.freebsd.org/builds?jailname=131amd64&type=package
> 2. http://beefy14.nyi.freebsd.org/build.html?mastername=131amd64-quarterly&build=07882fc18e19 for quarterly
> 3. `firebird30-client-3.0.7_2` amongst the failures
> 4. open _coredump_ in a new background tab
> 5. under _origin_, open _databases/firebird30-client_
> 6. edit `www` to `test` then add `#packages` to the tail of `https://test.freshports.org/databases/firebird30-client/`
> 7. at https://test.freshports.org/databases/firebird30-client/#packages click point at, but do not click, the quarterly (second) fallout icon for `FreeBSD:13:amd64`
> 8. from what's in the pop-up, memorise the build date
> 9. use the quarterly (second) fallout icon for `FreeBSD:13:amd64` any fallout icon to open https://portsfallout.com/fallout?port=databases%2Ffirebird30-client%24 in a new foreground tab
> 10. **ignore** the uppermost log for _131amd64-quarterly_
> 11. click the log for _131amd64-quarterly_ that matches the **memorised** build date
> 12. bring the other new tab (from _coredump_) to front
> 13. check the URLs of the two tabs.
> 
> The two are effectively the same:
> 
> * `http://beefy14.nyi.freebsd.org/data/131amd64-quarterly/07882fc18e19/logs/errors/firebird30-client-3.0.7_2.log` via the _coredump_ link
> * `http://beefy14.nyi.freebsd.org/data/131amd64-quarterly/07882fc18e19/logs/firebird30-client-3.0.7_2.log` – without `errors/` in the URL – via portsfallout.com
> 
> – and at https://lists.freebsd.org/archives/freebsd-pkg-fallout/2022-October/date.html by date for this month, the penultimate match for _131amd64-quarterly][databases/firebird30-client] Failed for firebird30-client_ is https://lists.freebsd.org/archives/freebsd-pkg-fallout/2022-October/279283.html, Sun, 09 Oct 2022 01:34:57 UTC.
> 


When I start reading this, I wonder:  what is this list? Why should  I look at this link? Holy shit, I have a lot to review here.  What?

Then I read:

> Test result: OK. **However** …

OK, no action required here. Phew.

**Comment by dlangille** [comment]
> ### UX
> … at https://test.freshports.org/databases/firebird30-client/#packages there's an implicit requirement to pay careful attention to the build date, if the end user is to identify the correct log file.
> 
> Let's be realistic about the likelihood of users actually paying attention and then knowing what to do with the date.
> 
> What's most obvious (uppermost) at the fallout list might be **not** an explanation for a gap in the table of packages at FreshPorts.
> 
> With the databases/firebird30-client example above, at the time of writing there's a parallel build at https://pkg-status.freebsd.org/beefy14/ in progress (neither done, nor crashed), it's this build that produced the log for 131amd64-quarterly that's currently uppermost at the fallout list.

My interpretation:  No action required.

**Comment by dlangille** [comment]
> https://test.freshports.org/graphics/py-cairo/ now has 275 fallout links to a single page that confirms, no fallout.
> 
> For a port that truly does have fallout: https://test.freshports.org/databases/firebird30-client/#packages **in theory**, you might customise each fallout cell within the table, to _precisely_ directly link to the historic reason for fallout (or skip) at that point in time.

How would that be done? For background, the lists which appear under Packages are meant to be a list of what you can download now. Every hour, at the top of the hour, the package repos are queried to see if there are any new builds available. The `packagesite.yaml` file is parsed and used to populate the database.  This blog post has more: https://news.freshports.org/2020/04/10/what-packages-are-now-available-for-this-port/

How does the code detect the right link to provide?

> **In reality**, for now, the more that I think about this, the _general_ direction to the (one) fallout page for a port is better served by the (one) existing fallout icon that's already at the head of production service (www).
> 
> Sorry … that's not intended to be defeatist. Just realistic about what's gained if multiple fallout links to a single page are added. (Am I missing something?)

The whole point of this issue was a link to the fallout page. They are all identical links right now. I'm happy to provide more specific links - perhaps getting the ABI values involved...

**Comment by grahamperrin** [comment]
> Phew

Yeah, I did want to use `Details` markup, but the result is messy. Done now: https://github.com/FreshPorts/freshports/issues/384#issuecomment-1275568706

**Comment by grahamperrin** [comment]
> … They are all identical links right now. I'm happy to provide more specific links …

Oh, that'll be smart, if it's possible. 

(I imagined it not possible. Sorry.)

**Comment by dlangille** [comment]
> > … They are all identical links right now. I'm happy to provide more specific links …
> 
> Oh, that'll be smart, if it's possible.
> 
> (I imagined it not possible. Sorry.)

I have not imagined how to do it. I will leave that to someone with more time. :)
